ursine/header.py

```diff
@@ -0,0 +1,106 @@
+import copy
+import random
+import string
+from .header_parsing import parse_header
+
+
+def random_tag():
+    return ''.join([random.choice(string.hexdigits) for _ in range(16)])
+
+
+class HeaderError(Exception):
+    pass
+
+
+class Header:
+    __slots__ = (
+        '_display_name',
+        '_parameters',
+        '_uri',
+    )
+
+    def __init__(self, header):
+        result = parse_header(header)
+        self._display_name = result.display_name
+        self._parameters = result.parameters
+        self._uri = result.uri
+
+    @classmethod
+    def build(cls, *, uri, display_name=None, parameters=None):
+        '''Build a new Header from kwargs.'''
+        self = object.__new__(cls)
+        self._uri = uri
+        self._display_name = display_name
+        self._parameters = parameters if parameters else {}
+        self._validate()
+        return self
+
+    display_name = property(lambda self: self._display_name)
+    parameters = property(lambda self: self._parameters)
+    uri = property(lambda self: self._uri)
+    tag = property(lambda self: self._parameters.get('tag', None))
+
+    def with_display_name(self, display_name):
+        new = copy.deepcopy(self)
+        new._display_name = display_name
+        new._validate()
+        return new
+
+    def with_uri(self, uri):
+        new = copy.deepcopy(self)
+        new._uri = uri
+        new._validate()
+        return new
+
+    def with_parameters(self, parameters):
+        new = copy.deepcopy(self)
+        new._parameters = parameters
+        new._validate()
+        return new
+
+    def with_tag(self, tag):
+        if self.tag == tag:
+            return self
+        new = copy.deepcopy(self)
+        if tag:
+            new._parameters['tag'] = tag
+        elif 'tag' in new._parameters:
+            del new._parameters['tag']
+        return new
+
+    def with_default_tag(self, tag=None):
+        '''Like `with_tag`, but will only alter
+        the tag if one is not already defined.'''
+        if self.tag:
+            return self
+        if not tag:
+            tag = random_tag()
+        return self.with_tag(tag)
```

Comment:
You don't need two functions for this. `with_tag()` should be totally valid.

Reply:
The distinction I'm making is between `with_tag(None)` meaning 'assign a `None` tag (IE delete any existing tag) whereas `with_default` will either set a given tag if no tag exists on the original or `with_default_tag(None)` will assign a randomly generated tag if no tag exists on the original.

If we just want to drop tag generation entirely and leave that up to users I can see that, but otherwise I think `with_tag` should follow what all the other methods do and use whatever you give it - including `None` - as the new value but something else like `with_default_tag` (Or I considered naming it `ensure_has_tag`) allows for adding a random tag.